src/irmin-pack/unix/file_manager.ml

```diff
@@ -0,0 +1,256 @@
+    (Dict : Append_only_file.S with module Io = Control.Io)
+    (Suffix : Append_only_file.S with module Io = Control.Io)
+    (Index : Pack_index.S) =
+struct
+  module Io = Control.Io
+
+  type t = {
+    dict : Dict.t;
+    control : Control.t;
+    suffix : Suffix.t;
+    index : Index.t;
+    use_fsync : bool;
+  }
+
+  let dict_is_about_to_auto_flush _ = ()
+  let suffix_is_about_to_auto_flush _ = ()
+  let index_is_about_to_auto_flush _ = ()
+
+  (* File creation ********************************************************** *)
+
+  let finish_constructing_rw config control ~make_dict ~make_suffix ~make_index
+      =
+    let open Result_syntax in
+    let root = Irmin_pack.Conf.root config in
+    let use_fsync = Irmin_pack.Conf.use_fsync config in
+    let pl : Payload.t = Control.payload control in
+    let generation =
+      match pl.status with
+      | From_v1_v2_post_upgrade _ | From_v3_gc_disallowed -> 0
+      | From_v3_gc_allowed x -> x.gc_generation
+    in
+    (* 1. Create a ref for dependency injections for auto flushes *)
+    let instance = ref None in
+    let get_instance () =
+      match !instance with None -> assert false | Some x -> x
+    in
+    (* 2. Open the other files *)
+    let* suffix =
+      let path = Irmin_pack.Layout.V3.suffix ~root ~generation in
+      let auto_flush_threshold =
+        Irmin_pack.Conf.suffix_auto_flush_threshold config
+      in
+      let cb () = suffix_is_about_to_auto_flush (get_instance ()) in
+      make_suffix ~path ~auto_flush_threshold ~auto_flush_callback:cb
+    in
+    let* dict =
+      let path = Irmin_pack.Layout.V3.dict ~root in
+      let auto_flush_threshold =
+        Irmin_pack.Conf.dict_auto_flush_threshold config
+      in
+      let cb () = dict_is_about_to_auto_flush (get_instance ()) in
+      make_dict ~path ~auto_flush_threshold ~auto_flush_callback:cb
+    in
+    let* index =
+      let log_size = Conf.index_log_size config in
+      let throttle = Conf.merge_throttle config in
+      let cb () = index_is_about_to_auto_flush (get_instance ()) in
+      (* [cb] will not be called during calls to [index.flush] because we will
+         use [~no_callback:()] *)
+      make_index ~flush_callback:cb ~readonly:false ~throttle ~log_size root
+    in
+
+    Ok { dict; control; suffix; use_fsync; index }
+
+  let create_control_file ~overwrite config (status : Payload.status) =
+    let root = Irmin_pack.Conf.root config in
+    let use_fsync = Irmin_pack.Conf.use_fsync config in
+    let path = Irmin_pack.Layout.V3.control ~root in
+    let pl =
+      let open Payload in
+      let z = Int63.zero in
+      { dict_offset_end = z; entry_offset_suffix_end = z; status }
+    in
+    Control.create_rw ~path ~overwrite ~use_fsync pl
+
+  (** Note on SWMR consistency: It is undefined for a reader to attempt an
+      opening before [create_rw] is over.
+
+      Note on crash consistency: Crashing during [create_rw] leaves the storage
+      in an undefined state.
+
```

Comment:
Can we do better than "undefined state" here?

Reply:
Yes we can, but I this it's OK to not worry about this now